Tests/NIOHTTP1Tests/HTTPServerPipelineHandlerTest.swift

```diff
@@ -0,0 +1,358 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+//
+//  HTTPServerPipelineHandlerTest.swift
+//  NIOHTTP1Tests
+//
+//  Created by Cory Benfield on 02/03/2018.
+//
+
+import XCTest
+import NIO
+import NIOHTTP1
+
+private final class ReadRecorder: ChannelInboundHandler {
+    typealias InboundIn = HTTPServerRequestPart
+
+    enum Event: Equatable {
+        case read(InboundIn)
+        case halfClose
+
+        static func ==(lhs: Event, rhs: Event) -> Bool {
+            switch (lhs, rhs) {
+            case (.read(let b1), .read(let b2)):
+                return b1 == b2
+            case (.halfClose, .halfClose):
+                return true
+            default:
+                return false
+            }
+        }
+    }
+
+    public var reads: [Event] = []
```

Comment:
should this really be publicly mutable?

Reply:
In the tests I don't think it matters.

Nope, didn't run perf tests, will do shortly.

Reply:
sorry, missed that it was in the tests, please ignore